### ISSUE TITLE
TorchServe Parallel Neuron Core Usage

### ISSUE THREAD

**Issue by ed-muthiah** [aws/aws-neuron-sdk#265]
Hello Neuron SDK team,

**TLDR:**
- Are inference workloads automatically distributed among neuron cores? If not, how can I do this?
- How do NeuronCore Groups interact with TorchServe workers and the handler?

**Context:**
I'm using TorchServe to deploy the an obect detector model on AWS and running predictions online.

**Setup:**
My server: Inf1.xlarge -> Runs TorchServe and the [object detector handler](https://github.com/pytorch/serve/blob/master/ts/torch_handler/object_detector.py)
My client: c4large -> Runs my inference application and gets predictions using `result= requests.put(TorchServeURL, image)`

**Question:**
Currently, I'm running only 1 client and 1 server. Because the object detector handler is a python script, I'm assuming that it can only run on a single Inferentia Neuron core. As I understand there are 4 total available Neuron Cores. I would like to have 10-100 unique client applications all pointing to the same server. How do I ensure that all 4 cores are used? Do I have to use Python's multiprocessing package somewhere?

From the [BERT TorchServe Tutorial](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-frameworks/pytorch-neuron/tutorials/tutorial-torchserve.html):
```
INITIAL_WORKERS=4 # number of models that will be loaded at launch
curl -X POST "http://localhost:8081/models?url=$MODEL_NAME.mar&batch_size=$BATCH_SIZE&initial_workers=$INITIAL_WORKERS&max_batch_delay=$MAX_BATCH_DELAY"
```
Returns `"status": "Model \"bert-max_length128-batch_size6\" Version: 1.0 registered with 4 initial workers"`

Then @aws-mesharma writes: _"TorchServe will instantiate the number of model handlers indicated by INITIAL_WORKERS, so this value controls how many models we will load onto Inferentia in parallel. This tutorial was performed on an inf1.xlarge instance (one Inferentia chip), so there are four NeuronCores available. If you want to control worker scaling more dynamically, see the docs."_

Does this mean that all that's required to automatically scale the number of neuron cores is just to specify `initial_workers=4` when initialising torchserve?

Another key question is how the TorchServe number of workers interacts with the concept of [NeuronCore Groups](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-frameworks/tensorflow-neuron/tutorials/tutorial-tensorflow-NeuronCore-Group.html). For example if I compile 2 parallel FastRCNN models onto 1 core each, and call them in the handler, does this mean that my max TorchServe workers is only 2? Because each TorchServe worker calls 2 models on 2 cores?

**Comment by aws-stdun** [comment]
Hi @ed-muthiah,

In this case, adjusting `INITIAL_WORKERS` will adjust the number of NeuronCores in use.
If you look at [handler_bert.py](https://awsdocs-neuron-staging.readthedocs-hosted.com/en/torchserve-update/neuron-guide/neuron-frameworks/pytorch-neuron/tutorials/tutorial-torchserve.html#id2), you will notice these lines:
```
# one core per worker
os.environ['NEURONCORE_GROUP_SIZES'] = '1'
```

To answer your second question, the line above specifies how TorchServe interacts with NeuronCore Groups. In this case, we have specified that the number of NeuronCores is simply 1 core / worker.

There are two things you should keep in mind: 
- I see that your setup includes an inf1.xlarge which only has 4 NeuronCores, so you will probably want a larger instance size to specify more than four. 
- The benchmarking code in that tutorial will work with more than 4 NeuronCores, but you may find better performance by tweaking the parameters further.

**Comment by awsrjh** [comment]
closing - please let us know if you need any further help
